### ISSUE TITLE
Target-specific ABIs are allowed in compatible targets

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#65443]
For example, one can use the `ptx-kernel` ABI on a x86_64 target. What the following does, I have absolutely no idea:

```rust
#![feature(abi_ptx)]

pub extern "ptx-kernel" fn foo(x: i32) -> i32 { x } 
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9b19b79167e8a2047593a2751609c70f))

Errors:

```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.53s

```

cc @eddyb @rkruppe

**Comment by gnzlbg** [comment]
@jonas-schievink I'm not sure what the best label would be, but this isn't just an FFI issue, since one can call `extern "ptx-kernel"` functions defined in Rust from Rust. Basically, `extern "ABI"` just lets you tell the Rust compiler which calling convention should it use for a function (well, kind of, because of implicit no-mangle in some contexts, but that's another story).

I think that we do not check for any of the ABIs whether they make sense for the target, so it is very likely that there is a stable ABI that can be used in a stable target for which it doesn't make much sense. 

For example, [the win32 ABI "stdcall" can be used on stable Rust on x86_64-unknown-linux-gnu. ](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=acf43017f39c6088c92f1b35c9b82a62). This does something, but I'm not sure what that something would be.

**Comment by jonas-schievink** [comment]
@gnzlbg Yeah, specifying an ABI is most commonly used for FFI though, and that was the closest label we have, so I picked that. Feel free to change it to something else.

**Comment by Centril** [comment]
@jonas-schievink btw, if the current label set is insufficient, feel free to add new ones. :)

**Comment by nagisa** [comment]
Similar to https://github.com/rust-lang/rust/issues/57182.
